### PR TITLE
Support for input vectors bigger that 2GB

### PR DESCRIPTION
**Pull request by mateuszpn**
mhp::copy and mhp::sort updated to avoid 32-bit int limitations

### CODE REVIEW

File: include/dr/detail/communicator.hpp [hunk 1]
```diff
@@ -227,7 +228,7 @@ class rma_window {
   void create(communicator comm, void *data, std::size_t size) {
     local_data_ = data;
     communicator_ = comm;
-    DRLOG("win create:: size: {} data:{}", size, data);
+    DRLOG("win create:: size: {} data:{}", (MPI_Aint)size, data);
```
Comment (lslusarczyk) [inline include/dr/detail/communicator.hpp:231]: 1. no c-style casts
2. why the cast is needed?

File: include/dr/detail/communicator.hpp [hunk 2]
```diff
@@ -264,12 +273,23 @@ class rma_window {
 
   void put(const void *src, std::size_t size, std::size_t rank,
            std::size_t disp) const {
-    DRLOG("MPI comm put:: ({}:{}:{})", rank, disp, size);
+    DRLOG("MPI comm put:: ({}:{}:{})", rank, disp, (int)size);
```
Comment (lslusarczyk) [inline include/dr/detail/communicator.hpp:276]: why cast?

File: include/dr/detail/communicator.hpp [hunk 3]
```diff
@@ -264,12 +273,23 @@ class rma_window {
...
     MPI_Request request;
+
+#if (MPI_VERSION >= 4) ||                                                      \
```
Comment (lslusarczyk) [inline include/dr/detail/communicator.hpp:279]: to not repeat long condition please create: `#define MPI_SUPPORTS_RGET_C ...` and use in both places

File: include/dr/detail/communicator.hpp [hunk 4]
```diff
@@ -264,12 +273,23 @@ class rma_window {
...
+#else
+    // MPI_Rput origin_count is 32-bit signed int - check range
+    assert(
+        size <= (std::size_t)INT_MAX &&
+        "MPI API requires origin_count to be positive signed 32-bit integer");
     MPI_Rput(src, size, MPI_BYTE, rank, disp, size, MPI_BYTE, win_, &request);
-    DRLOG("MPI comm wait:: ({}:{}:{})", rank, disp, size);
+#endif
+
+    DRLOG("MPI comm wait:: ({}:{}:{})", rank, disp, (MPI_Aint)size);
```
Comment (lslusarczyk) [inline include/dr/detail/communicator.hpp:290]: casts

File: include/dr/mhp/containers/distributed_vector.hpp [hunk 5]
```diff
@@ -38,15 +38,43 @@ class MpiBackend {
...
     DRLOG("calling MPI put(segm_offset:{}, "
           "src:{}, size:{}, peer:{})",
           offset, src, datalen, segment_index);
+
+#if (MPI_VERSION >= 4) ||                                                      \
+    (defined(I_MPI_NUMVERSION) && (I_MPI_NUMVERSION > 20211200000))
+    // 64-bit API inside
     win_.put(src, datalen, segment_index, offset);
+#else
+    for (std::size_t remainder = datalen, off = 0UL; remainder > 0;) {
```
Comment (lslusarczyk) [inline include/dr/mhp/containers/distributed_vector.hpp:68]: helper function to not repeat code in two places?

File: include/dr/detail/communicator.hpp (hunk 1 above)
Comment (mateuszpn) [reply include/dr/detail/communicator.hpp:231]: It was necessary during debug (usigned/signed conversion), now removed,

File: include/dr/mhp/containers/distributed_vector.hpp (hunk 5 above)
Comment (mateuszpn) [reply include/dr/mhp/containers/distributed_vector.hpp:68]: The difference in the middle (get/put) prevents it. 
In fact, when IMPI 2021.12 will be officially released, and older versions removed from our tests, the code for older MPI should be removed.

**Pull request merged**